src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -102,13 +111,228 @@ private static async Task<VersionStamp> GetDocumentSemanticVersionAsync(Document
             return semanticVersion;
         }
 
-        // Internal for testing
-        internal async Task<SemanticTokens?> GetSemanticTokensAsync(
+        private async Task<SemanticTokens?> GetCachedTokensAsync(
             TextDocumentIdentifier textDocumentIdentifier,
             DocumentSnapshot documentSnapshot,
             int documentVersion,
             VersionStamp semanticVersion,
             Range range,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            if (!_cachedResponses.TryGetValue(textDocumentIdentifier.Uri, out var semanticVersionCache) ||
+                !semanticVersionCache.TryGetValue(semanticVersion, out var documentCache))
+            {
+                return null;
+            }
+
+            // 1) We'll first check if the cache contains an exact match for the range, in which case we can
+            // return early.
+            if (documentCache.TryGetValue(range, out var cachedTokens))
+            {
+                return cachedTokens;
+            }
+
+            // 2) If the cache doesn't contain an exact match for the range, we can still check if there's a
+            // partial match. This allows us to reduce the range we have to compute tokens for. We want to find
+            // the cached range with the most overlap with the requested range.
+            (Range Range, int NumOverlappingLines)? currentBestMatch = null;
+            foreach (var rangeCandidate in documentCache.Keys)
+            {
+                if (!range.OverlapsWith(rangeCandidate))
+                {
+                    continue;
+                }
+
+                var overlap = range.Overlap(rangeCandidate);
+                Assumes.NotNull(overlap);
+
+                // It's possible that the cached range candidate encompasses a span that completely overlaps with
+                // the requested range. Nothing in the LSP spec explicitly prevents us from returning excess
+                // tokens outside of the requested range, so if we hit this case we can return early.
+                if (overlap.Equals(range) && documentCache.TryGetValue(rangeCandidate, out cachedTokens))
+                {
+                    return cachedTokens;
+                }
+
+                // The range candidate becomes the best range either if:
+                //     1) It is the first candidate we are analyzing
+                //     2) The # of overlapping lines exceeds the current best candidate, *and* the current candidate
+                //     is not in the middle of the requested range. (We avoid the latter since it complicates later
+                //     computations.)
+                var numOverlappingLines = overlap.End.Line - overlap.Start.Line;
+                if (currentBestMatch is null ||
+                    ((currentBestMatch.Value.NumOverlappingLines < numOverlappingLines) && (rangeCandidate.Start <= range.Start || rangeCandidate.End >= range.End)))
```

Comment:
Instead of (`(rangeCandidate.Start <= range.Start || rangeCandidate.End >= range.End)`) would reversing the overlaps function have the same effect? Aka `rangeCandidate.OverlapsWith(range)`